pysrc/bytewax/connectors/kafka/operators.py

```diff
@@ -0,0 +1,289 @@
+        A stream of `KafkaMessage`s where `KafkaMessage.error is None`
+    - KafkaStreams.errors:
+        A stream of `KafkaMessage`s where `KafkaMessage.error is not None`
+    """
+
+    oks: Stream[KafkaMessage[K, V]]
+    errs: Stream[KafkaMessage[K2, V2]]
+
+
+@operator
+def _kafka_error_split(
+    step_id: str, up: Stream[Union[KafkaMessage[K, V], KafkaMessage[K2, V2]]]
+) -> KafkaStreams[K, V, K2, V2]:
+    """Split a stream of KafkaMessages in two."""
+
+    def predicate(msg: KafkaMessage) -> bool:
+        return msg.error is None
+
+    branch = op.branch("branch", up, predicate)
+    # Cast the streams to the proper expected types.
+    # This assumes that messages where error is None will be KafkaMessage[K, V]
+    # while messages where error is not None will be KafkaMessage[K2, V2]
+    oks = cast(Stream[KafkaMessage[K, V]], branch.trues)
+    errs = cast(Stream[KafkaMessage[K2, V2]], branch.falses)
+    return KafkaStreams(oks, errs)
+
+
+_default_add_config: Dict = {}
+
+
+@operator
+def input(  # noqa A001
+    step_id: str,
+    flow: Dataflow,
+    *,
+    brokers: List[str],
+    topics: List[str],
+    tail: bool = True,
+    starting_offset: int = OFFSET_BEGINNING,
+    add_config: Optional[Dict[str, str]] = None,
+    batch_size: int = 1000,
+) -> KafkaStreams[MaybeStrBytes, MaybeStrBytes, MaybeStrBytes, MaybeStrBytes]:
+    """Use a set of Kafka topics as an input source.
+
+    Partitions are the unit of parallelism.
+    Can support exactly-once processing.
+
+    Messages are emitted into the dataflow
+    as `bytewax.connectors.kafka.KafkaMessage` objects.
+
+    Args:
+        step_id:
+            Unique name for this step
+        flow:
+            A Dataflow
+        brokers:
+            List of `host:port` strings of Kafka brokers.
+        topics:
+            List of topics to consume from.
+        tail:
+            Whether to wait for new data on this topic when the
+            end is initially reached.
+        starting_offset:
+            Can be either `confluent_kafka.OFFSET_BEGINNING` or
+            `confluent_kafka.OFFSET_END`. Defaults to beginning of
+            topic.
+        add_config:
+            Any additional configuration properties. See the `rdkafka`
+            [docs](https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md)
+            for options.
+        batch_size:
+            How many messages to consume at most at each poll. Defaults to 1000.
+    """
+    return op.input(
+        "kafka_input",
+        flow,
+        KafkaSource(
+            brokers,
+            topics,
+            tail,
+            starting_offset,
+            add_config,
+            batch_size,
+            # Don't raise on errors, since we split
+            # the stream and let users handle that
+            raise_on_errors=False,
+        ),
+    ).then(_kafka_error_split, "split_err")
+
+
+@operator
+def output(
+    step_id: str,
+    up: Stream[Tuple[bytes, bytes]],
```

Comment:
Yes I did notice the same when writing the dataflow for the kafka-to-kafka benchmark, and didn't like that I had to add a map step in between.

I left the output stream as it was because I didn't want to force the use of the class when writing to kafka from a different source, but I agree, since we can specify more metadata when producing too, it makes sense to have the same ability there.

As you noted, some fields can be ignored on the output, but others like `topic`, `partition` and `timestamp` have different meanings in the source and sink. I'd like to differentiate the two things, because if we also allow to select a specific topic for each message, you could end up with an output that automatically produces in the input topic (infinite data yeee).

I think we have two ways to do that:

1) "Unfreeze" `KafkaMessage` and add the fields `out_topic`, `out_timestamp`, `out_partition` to the class, and use it in both source and sink
2) Use 2 different classes, but accept both in the output. 

In case `2` we'd have `KafkaSourceMessage` and `KafkaSinkMessage`.
`KafkaSourceMessage` has a method to safely convert to `KafkaSinkMessage` by only reusing `key`, `value` and `headers`. In the output and serialize* operators, `KafkaSourceMessage` will be automatically converted to `KafkaSinkMessage`.

Then we can make it so that if you explicitly pass `topic=None` to the sink, the topic is read from each message (and it has to be `KafkaSinkMessage` at that point).

With both options you need to explicitly set the out topic, either by setting `out_topic`, or by converting a `KafkaSourceMessage` to a `KafkaSinkMessage`.

I'll give a try to the `2` option, but let me know what you think

Reply:
So, we end up with:
```mermaid
graph TD;
kafka.input -- "KafkaSourceMessage[MaybeStrBytes, MaybeStrBytes]" --> kafka.output;
```
```mermaid
graph TD;
kafka.input -- "KafkaSourceMessage[MaybeStrBytes, MaybeStrBytes]" --> kafka.deserialize -- "KafkaSourceMessage[K, V]" --> kafka.serialize -- "KafkaSinkMessage[bytes, bytes]" --> kafka.output;
```

You can always do an explicit `map` step to create the `KafkaSinkMessage`, but if you don't, you get the automatic conversion in `kop.output` and `kop.serialize*`. If you use the sink directly, it's always up to you to do the conversion.